processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -133,84 +188,37 @@ public int compare(String o1, String o2)
     int valuesOffset = theBuffer.position() + size * Ints.BYTES;
 
     buffer.position(valuesOffset);
-    valueBuffers = Lists.newArrayList(buffer.slice());
+    // Ensure the value buffer's limit equals to capacity.
+    firstValueBuffer = buffer.slice();
+    valueBuffers = new ByteBuffer[]{firstValueBuffer};
     buffer.position(indexOffset);
     headerBuffer = buffer.slice();
-    final ByteBuffer valueBuffer = valueBuffers.get(0);
-    bufferIndexed = new BufferIndexed()
-    {
-      @Override
-      public T get(int index)
-      {
-        checkIndex(index, size);
-
-        final int startOffset;
-        final int endOffset;
-
-        if (index == 0) {
-          startOffset = 4;
-          endOffset = headerBuffer.getInt(0);
-        } else {
-          int headerPosition = (index - 1) * Ints.BYTES;
-          startOffset = headerBuffer.getInt(headerPosition) + Ints.BYTES;
-          endOffset = headerBuffer.getInt(headerPosition + Ints.BYTES);
-        }
-        return _get(valueBuffer.asReadOnlyBuffer(), startOffset, endOffset);
-      }
-    };
   }
 
-  // used for multiple file version, v2.
+
+  /**
+   * Constructor for version two.
+   */
   GenericIndexed(
-      List<ByteBuffer> valueBuffs,
+      ByteBuffer[] valueBuffs,
       ByteBuffer headerBuff,
       ObjectStrategy<T> strategy,
       boolean allowReverseLookup,
       int logBaseTwoOfElementsPerValueFile,
       int numWritten
   )
   {
+    this.versionOne = false;
```

Comment:
I agree about separate (redundant) field, if the code like `buffer.size() == 1` is extracted as a method is `isVersionOne()`.

This field is used in `get()`: `return versionOne ? getVersionOne(index) : getVersionTwo(index);`, the only anywhere hot method in `GenericIndexed`.